js/widgets/musickeyboard.js

```diff
@@ -1,73 +1,270 @@
+ *
+ * @license
+ * This program is free software; you can redistribute it and/or modify it under the terms of the
+ * The GNU Affero General Public License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * You should have received a copy of the GNU Affero General Public License along with this
+ * library; if not, write to the Free Software Foundation, 51 Franklin Street, Suite 500 Boston,
+ * MA 02110-1335 USA.
+ */
+
+/*
+   global platformColor, docById, wheelnav, slicePath, logo, Singer, last, _, FIXEDSOLFEGE1
+   SHARP, FLAT, noteToFrequency, EIGHTHNOTEWIDTH, MATRIXSOLFEHEIGHT, MATRIXSOLFEWIDTH,
+   i18nSolfege, toFraction, blocks, getNote, PREVIEWVOLUME, DEFAULTVOICE, PITCHES3,
+   SOLFEGENAMES, PITCHES2, PITCHES, NOTESSHARP, NOTESFLAT, convertFromSolfege, SOLFEGECONVERSIONTABLE
+
+
+ */
+
+/* exported MusicKeyboard */
+
+/**
+ * @class
+ * @classdesc pertains to setting up all features of the musickeyboard widget and its UI features.
+ *
+ * Private members' names begin with underscore '_".
+ */
+class MusicKeyboard {
+    static FAKEBLOCKNUMBER = 100000;
+    static BUTTONDIVWIDTH = 535; // 5 buttons
+    static OUTERWINDOWWIDTH = 758;
+    static INNERWINDOWWIDTH = 50;
+    static BUTTONSIZE = 53;
+    static ICONSIZE = 32;
     // Mapping between keycodes and virtual keyboard
-    const BLACKKEYS = [87, 69, 82, 84, 89, 85, 73, 79];
-    const WHITEKEYS = [65, 83, 68, 70, 71, 72, 74, 75, 76];
-    const SPACE = 32;
+    static BLACKKEYS = [87, 69, 82, 84, 89, 85, 73, 79];
+    static WHITEKEYS = [65, 83, 68, 70, 71, 72, 74, 75, 76];
+    static SPACE = 32;
 
-    const saveOnKeyDown = document.onkeydown;
-    const saveOnKeyUp = document.onkeyup;
+    static saveOnKeyDown = document.onkeydown;
+    static saveOnKeyUp = document.onkeyup;
+    static beginnerMode = localStorage.beginnerMode;
+    static unit = MusicKeyboard.beginnerMode === "true" ? 8 : 16;
+    static selectedNotes = [];
 
-    const w = window.innerWidth;
-    this._cellScale = w / 1200;
+    /**
+    * @constructor
+    */
+
+    constructor() {
+        const w = window.innerWidth;
+        this._cellScale = w / 1200;
+
+        this._stopOrCloseClicked = false;
+        this.playingNow = false;
+
+        this.instruments = [];
+        this.noteNames = [];
+        this.octaves = [];
+        this.keyboardShown = true;
+        this.layout = [];
+        this.idContainer = [];
+        this.tick = false;
+        this.meterArgs = [4, 1 / 4];
+
+        // Map between keyboard element ids and the note associated with the key.
+        this.noteMapper = {};
+        this.blockNumberMapper = {};
+        this.instrumentMapper = {};
+
+        this._rowBlocks = [];
+
+        // Each element in the array is [start time, note, id, duration, voice].
+        this._notesPlayed = [];
+    }
+
+    /**
+     * Initialises the MusicKeyboard widget.
+     * @returns {void}
+     */
+
+    init() {
```

Comment:
would be better if we could remove the `init` function and do things in the constructor instead, please make the necessary changes where the class is instantiated as well. We could make another function to setup the UI elements and event listeners and call it from the constructor for making our code more organised. Thanks

Reply:
It's imperative to reach a consensus on this. 

- phrasemaker
- pitchdrummatrix
- pitchslider
- pitchstaircase
- rhythmruler
- status
- temperament 
- tempo
- timbre

all of these widgets have `init` function. If we're deciding to eradicate `init` function all together, it be done with all widgets and not just arbitrarily.